python/ray/data/_internal/execution/operators/zip_operator.py

```diff
@@ -236,11 +237,32 @@ def _zip_one_block(
         # Swap blocks if ordering was inverted during block alignment splitting.
         block, other_block = other_block, block
     # Zip block and other blocks.
-    result = BlockAccessor.for_block(block).zip(other_block)
+    result = _try_zip_with_pyarrow_fallback(block, other_block)
     br = BlockAccessor.for_block(result)
     return result, br.get_metadata(input_files=[], exec_stats=stats.build())
 
 
 def _get_num_rows_and_bytes(block: Block) -> Tuple[int, int]:
     block = BlockAccessor.for_block(block)
     return block.num_rows(), block.size_bytes()
+
+
+def _try_zip_with_pyarrow_fallback(block: Block, other_block: Block) -> Block:
+    """Try zipping blocks together. If zip fails on ValueError exception due to
+    mismatched block formats we convert `block` or `other_block` to arrow format
+    and zip again.
+    """
+    try:
+        return BlockAccessor.for_block(block).zip(other_block)
+    except ValueError as e:
+        pyarrow_table = _lazy_import_pyarrow_table()
```

Comment:
Why do we need `_lazy_import_pyarrow_table()` here? What would happen if we just did `from pyarrow import Table`?

Reply:
`from pyarrow import Table` works. The lazy import just reduces some import overhead when you call on the zip operator more than once.